test/functional/normal/undo_spec.lua

```diff
@@ -0,0 +1,58 @@
+local helpers = require('test.functional.helpers')(after_each)
+
+local clear = helpers.clear
+local expect = helpers.expect
+local feed = helpers.feed
+local insert = helpers.insert
+
+describe('undo/redo command', function()
+  local function create_history(num_steps)
+    if num_steps == 0 then return end
+    insert('1')
+    if num_steps == 1 then return end
+    feed('o2<esc>')
+    feed('o3<esc>')
+    feed('u')
+    if num_steps == 2 then return end
+    feed('o4<esc>')
+    if num_steps == 3 then return end
+    feed('u')
+  end
+
+  local function undo_and_redo(hist_pos, undo, redo, expect_str)
+    clear()
```

Comment:
I added `before_each(clear)` and changed this line to:

    execute('enew!')

That cut the test time in half (.05s vs .09s in local env).